R doc 1703 time series (#1152)
ravendb/docs: Documentation/5.0/Raven.Documentation.Pages/document-extensions/timeseries/design.markdown

﻿# Time-Series Design

---

{NOTE: }

> A time-series is an array of Gorilla-encoded `IEnumerable` entries 

Reviewer: This should probably be the second item on the menu

Reviewer 2: I moved this to its own section. Did I understand your intent correctly?

> A time-series is an array of Gorilla-encoded `IEnumerable` entries 

Reviewer: And you need first to explain what they _are_

Reviewer 2: Fixed

that extends a specific single document.  

* In this page:  
  * [Time-Series Structure](../../document-extensions/timeseries/design#time-series-structure)  
     * [Document Extension](../../document-extensions/timeseries/design#document-extension)  
     * [The `HasTimeSeries` Flag](../../document-extensions/timeseries/design#the--flag)  
     * [Segmentation](../../document-extensions/timeseries/design#segmentation)  
     * [Time-Series Entries](../../document-extensions/timeseries/design#time-series-entries)  
  * [Updating Time-Series](../../document-extensions/timeseries/design#updating-time-series)  
     * [Document Change](../../document-extensions/timeseries/design#document-change)  
     * [Success](../../document-extensions/timeseries/design#success)  
     * [No Conflicts](../../document-extensions/timeseries/design#no-conflicts)  
     * [Transactions](../../document-extensions/timeseries/design#transactions)  

{NOTE/}

---

{PANEL: Time-Series Structure}

---

#### Document Extension  

> A time-series is related to the document it extsnds, by -  

Reviewer 2: Fixed

* A **reference to the time-series** in the document's metadata.  
  The time-series' **name** is kept in the document's metadata.  

>   The time-series' **data** is stored in a different location.  '

Reviewer: ```suggestion
  The time-series' **data** is stored in a different location.  
```

Reviewer 2: Fixed

* A **reference to the document** in the time-series' data.  

---

#### The `HasTimeSeries` Flag

* When the first time-series is added to a document, RavenDB automatically sets 
  a `HasTimeSeries` Flag in the document's metadata.
  {CODE-BLOCK: JSON}
{
    "Name": "Paul",
    "@metadata": {
        "@collection": "Users",
        "@flags": "HasTimeSeries"
    }
}
{CODE-BLOCK/}

* When all time-series have been removed from a document, RavenDB will 
  automatically remove the flag.  

---

#### Segmentation

A time-series array is constructed of segments.  

> * A time-series of up to 32k entries is contained in a single segment.  

Reviewer: Also according to size. 
Important to stress that this is an impl. detail

Reviewer: Worth putting this here.

Reviewer 2: Done

> * A time-series longet than 32k entries is contained in multiple segments.  

Reviewer 2: fixed

* A new segment is also created when a period of more than 25 days has passed 
  since the last measurement.  

---

#### Time-Series Entries

> Each time-series entry is an `IEnumerable` with three values.  

Reviewer: What do you mean, IEnumerable? How does this matter?

Reviewer 2: fixed

* `TimeSeriesEntry` 

    | Parameters | Type | Description |
    |:-------------|:-------------|:-------------|

>     | `Timestamp` | DateTime | Time signature |

Reviewer: ```suggestion
    | `Timestamp` | DateTime UTC | The time of the event |
```

Reviewer 2: Fixed

>     | `Tag` | string | Optional tag |

Reviewer: Need to explain what the tag is

Reviewer 2: Elaborated

>     | `Values` | double[] | An array of up to 32 values |

Reviewer 2: Done

{PANEL/}

{PANEL: Updating Time-Series}

---

#### Document Change  

* **Name Change**  
  **Creating** or **removing** a time-series adds or removes its name 
  to/from the metadata of the document it extends.  
  These actions **will** invoke a document-change event.  
* **Data Updates**  
  Modifying time-series data does **not** invoke a document-change event, 
  as long as it doesn't create a new time-series or remove an existing one.  

---

#### Success

Updating a time-series is designed to succeed without causing any conflict.  
As long as the document it extends exists, updating a Time-Serie will always succeed.  

---

#### No Conflicts

Time-series actions do not cause conflicts.  

* **Time-series update by multiple cluster Nodes**  
   * When a time-series' data is replicated by multiple nodes, the data 
     from all nodes is merged into a single series.  
   * When multiple nodes append **different values** at the same timestamp, 

>      the **bigger value** is chosen for this entry.  

Reviewer: Not the bigger value. The entry that is sorted lexically higher is chosen.
Important because you may have multiple values.

Reviewer 2: fixed

>    * When multiple nodes apppend **different amount of values** for the same 

Reviewer: No necessarily. I think.
If the first value sort higher, it will pick that one, even if it has less values.
@karmeli87  ?

Reviewer: Okay, in this case, let's put it as such in the docs

Reviewer 2: Fixed

     timestamp, the **bigger amount of values** is chosen for this entry.  
   * When an existing value at a certain timestamp is deleted by a node 
     and updated by another node, the deletion is chosen.  

* **Time-series update By multiple clients of the same node**  
   * When a time-series' value at a certain timestamp is appended by 

>      multiple clients more or less simultaneously, the last one is chosen.  

Reviewer: ```suggestion
     multiple clients more or less simultaneously, RavenDB uses the last write strategy.
```

Reviewer 2: fixed

> ﻿# Time Series Design

Author: done

document and the time series reference each other through:  

* A **reference to the time series** in the document's metadata.  
  The time series' **name** is kept in the document's metadata.  
  The time series' **data** is stored in a separate location.  
* A **reference to the document** in the time series data.  

---

#### Time Series Entries

Each time series entry is composed of:  

* `TimeSeriesEntry` 

    | Parameters | Type | Description |
    |:-------------|:-------------|:-------------|
    | `Timestamp` | DateTime (UTC) | The time of the event or data represented by the entry. Time is measured up to millisecond resolution. |
    | `Tag` | string | An optional tag for an entry. Can be any string up to 255 bytes. Possible uses for the tag: descriptions or metadata for individual entries; storing a document id, which can then be referenced when querying a time series. This is the only component of the entry that is not numerical. |
    | `Values` | double[] | An array of up to 32 `double` values |

Doubles with higher precision - i.e. more digits after the decimal point - 
are much less compressible. In other words `1.234` takes up more space 

> than `1,234`.

Author: done

---

#### The `HasTimeSeries` Flag

* When a document has one or more time series, RavenDB automatically adds  
  a `HasTimeSeries` flag in the document's metadata under `@flags`:

{CODE-BLOCK: JSON}
{
    "Name": "Paul",
    "@metadata": {
        "@collection": "Users",
        "@timeseries": [
            "my time series"
        ]
        "@flags": "HasTimeSeries"
    }
}
{CODE-BLOCK/}

* When all time series are removed from a document, RavenDB 
automatically removes the flag.  

---

#### Segmentation

At the server storage level, time series data is divided into **segments**.  

* Segments contain a number of consecutive entries from the same 
time series.  
* Segments have a maximum size of 2 kB.  

> * The 2 kb limit means that a segment can only contain up to 32k entries,

Author: done

so a time series with more than 32k entries will always be stored in
multiple segments.  
* In practice, segments usually contain far fewer than 32k entries, 
the amount depends on the size of the entries (after compression).  
  * As an example - in the [Northwind sample dataset](../../studio/database/tasks/create-sample-data), 
  the _Companies_ documents all have a time series called _StockPrice_. 
  These time series are stored in segments that have ~10-20 entries each.  
* The maximum gap between the first and last entries in a segment is 
is ~24.86 days (`int.MaxValue` milliseconds). So if an entry is 24.86 days 
or later than the first entry in the current segment, it will become the 
first entry in a new segment. A consquence of this is that in a sparse time 

> series segments can be significantly smaller than 2 kB.  

Author: done

Segments also store the aggregate values that summarize the data they 
contain, such as the averages of entry values. For this reason, it's 
often useful to reference individual segments in indexes and queries. 
Read more about how to use segments [here](../../document-extensions/timeseries/indexing#object)

---

#### Compression

Time series data is stored using a format called [Gorilla compression](https://www.vldb.org/pvldb/vol8/p1816-teller.pdf). 
On top of the Gorilla compression, the time series segments are compressed 

> using the [Lz4 algorithm](https://lz4.github.io/lz4/).

Author: fixed to LZ4
